keps/sig-apps/3335-statefulset-slice/README.md

```diff
@@ -0,0 +1,907 @@
+*   Orchestrating pod movement from one StatefulSet slice to another
+*   Managing network connectivity between pods in different StatefulSet slices
+*   Orchestrating storage lifecycle of PVCs and PVs across different StatefulSet slices
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+This KEP solves the problem of managing subsets of the replicas in a StatefulSet by introducing the concept of a slice. A slice consists of a start ordinal `k`, and a number of replicas `N`. To control the starting and ending ordinal of each slice, a new field `replicaStartOrdinal` is introduced to `StatefulSetSpec`.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+The main motivation of this KEP is to support a more flexible StatefulSet, a building block in an ecosystem where Stateful applications can be migrated across Kubernetes clusters with more automation. Below are two high level user stories that share the problem of having a StatefulSet locked into a specific configuration. To fully automate these two scenarios, additional building blocks are needed around volume management and networking (see [Notes/Constraints/Caveats](#notesconstraintscaveats-optional)).
+
+#### Story 1
+
+**Migrating across namespaces**: Many organizations use namespaces for team isolation. Consider a team that is migrating a `StatefulSet` to a new namespace in a cluster. Migration could be motivated by a branding change, or a requirement to move out of a shared namespace. Consider the StatefulSet `my-app` with `replicas: 5`, running in a shared namespace.
+
+
+```
+name: my-app
+namespace: shared
+replicas: 5
+-----------------------------------------------
+[ nginx-0, nginx-1, nginx-2, nginx-3, nginx-4 ]
+```
+
+
+To move two pods, the `my-app` StatefulSet in the `shared` namespace can be scaled down to `replicas: 3, replicaStartOrdinal: 0`, and an analogous StatefulSet in the `app-team` namespace scaled up to `replicas: 2, replicaStartOrdinal: 3`. This allows for pod ordinals to be managed during migration. The application operator should manage network connectivity, volumes and slice orchestration (when to migrate and by how many replicas).
+
+
+```
+name: my-app						name: my-app
+namespace: shared					namespace: app-team
+replicas: 3						replicas: 2
+replicaStartOrdinal: 0				replicaStartOrdinal: 3
```

Comment:
So the idea behind slice orchestration is that the application operator will continuously update the values of `replicas` and `replicasStartOrdinal` until all the replicas are migrated over to the `app-team` namespace? It can also decide whether to continue or halt the rest of the migration based on e.g., new replicas' health etc.?

Reply:
Yes, that's correct. It provides the application operator granularity and control when moving replicas over.